flink-runtime/src/main/java/org/apache/flink/runtime/io/network/netty/BufferResponseAndNoDataBufferMessageParser.java

```diff
@@ -0,0 +1,60 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.runtime.io.network.netty;
+
+import org.apache.flink.shaded.netty4.io.netty.buffer.ByteBuf;
+
+/**
+ * The parser for both {@link NettyMessage.BufferResponse} and messages without data buffer.
+ */
+public class BufferResponseAndNoDataBufferMessageParser extends NoDataBufferMessageParser {
+
+    /** The allocator for the Flink buffer. */
+    private final NetworkBufferAllocator networkBufferAllocator;
+
+    BufferResponseAndNoDataBufferMessageParser(NetworkBufferAllocator networkBufferAllocator) {
+        this.networkBufferAllocator = networkBufferAllocator;
+    }
+
+    @Override
+    public int getMessageHeaderLength(int lengthWithoutFrameHeader, int msgId) {
+        if (msgId == NettyMessage.BufferResponse.ID) {
```

Comment:
It is tricky to let parser understand the implementation detail that `BufferResponse` has different header length with other messages. It is better to make this thing happen hide inside the `BufferResponse`.

Another possible option is to provide a static method to parse message header inside `BufferResponse`. Then after getting the msg id from frame header, `ZeroCopyNettyMessageDecoder` can pass the `messageHeaderBuffer` to the `BufferResponse` for parsing the message header. The `BufferResponse` is aware of the internal header length and the returned boolean tag can indicate whether the header is complete. Then we can remove the `getMessageHeaderLength` method from parser.

Reply:
I think it should be hard to achieve to let the `ZeroCopyNettyMessageDecoder` to decide whether the `messageHeaderBuffer` has been fulfilled without knowing its size in advance: When receiving a new piece of data from the socket, if we do not known the size of the message header part, we could not decide which part should be copied into `messageHeaderBuffer`. 

I also agree with that it is better to decide the `size` inside the messages. Thus currently I added a static `getMessageHeaderLength ` method to `ErrorResponse` and `BufferResponse` class and delegate the size acquisition to the specific messages, which is similar to the current pattern of `readFrom`.